nexus/src/db/datastore.rs

```diff
@@ -1112,6 +1135,272 @@ impl DataStore {
+                    disk::dsl::attach_instance_id.eq(authz_instance.id())
+                ))
+        )
+        .attach_and_get_result_async(self.pool_authorized(opctx).await?)
+        .await
+        .or_else(|e| {
+            match e {
+                AttachError::CollectionNotFound => {
+                    Err(Error::not_found_by_id(
+                        ResourceType::Instance,
+                        &authz_instance.id(),
+                    ))
+                },
+                AttachError::ResourceNotFound => {
+                    Err(Error::not_found_by_id(
+                        ResourceType::Disk,
+                        &authz_disk.id(),
+                    ))
+                },
+                AttachError::NoUpdate { attached_count, resource, collection } => {
+                    let disk_state = resource.state().into();
+                    match disk_state {
+                        // Idempotent errors: We did not perform an update,
+                        // because we're already in the process of attaching.
+                        api::external::DiskState::Attached(id) if id == authz_instance.id() => {
+                            return Ok((collection, resource));
+                        }
+                        api::external::DiskState::Attaching(id) if id == authz_instance.id() => {
+                            return Ok((collection, resource));
+                        }
+                        // Ok-to-attach disk states: Inspect the state to infer
+                        // why we did not attach.
+                        api::external::DiskState::Creating |
+                        api::external::DiskState::Detached => {
+                            match collection.runtime_state.state.state() {
+                                // Ok-to-be-attached instance states:
+                                api::external::InstanceState::Creating |
+                                api::external::InstanceState::Stopped => {
+                                    // The disk is ready to be attached, and the
+                                    // instance is ready to be attached. Perhaps
+                                    // we are at attachment capacity?
+                                    if attached_count == i64::from(max_disks) {
+                                        return Err(Error::invalid_request(&format!(
+                                            "cannot attach more than {} disks to instance",
+                                            max_disks
+                                        )));
+                                    }
+
+                                    // We can't attach, but the error hasn't
+                                    // helped us infer why.
+                                    return Err(Error::internal_error(
+                                        "cannot attach disk"
+                                    ));
+                                }
```

Comment:
(this is an example of the error not providing enough context, referenced in another review comment)

Reply:
I agree with this, but I'm still not sure what other info to attach. I think if this case happened, we'd need to perform manual inspection of the DB to infer why it happened.

I view this as akin to "attempting, but failing, to deserialize a value from the database". I don't think it's appropriate to dump raw strings all the way back to the client, but it's still worth flagging.